### ISSUE TITLE
[SWIFT] - MaplyShapeLinear

### ISSUE THREAD

**Issue by Sebboulnois** [mousebird/WhirlyGlobe#995]
Hello everyone,

I have been struggling with the creation of a MaplyShapeLinear in Swift.

Here is the class I made:

<img width="913" alt="screen shot 2017-09-04 at 12 41 32 pm" src="https://user-images.githubusercontent.com/15698398/30034323-9d93de8e-916e-11e7-88b0-fb07f1c43a45.png">

I basically fill up an array of MaplyCoordinate3D with the center properties of MaplyBillBoard. Since the MaplyShapeLinear requires an UnsafeMutablePointer during initialization, I turn my array of 3D coordinates into this type. At the end, I add this UnsafeMutablePointer variable in the addSapes function:

<img width="914" alt="screen shot 2017-09-04 at 12 41 51 pm" src="https://user-images.githubusercontent.com/15698398/30034389-04999f88-916f-11e7-92e8-ea6a4f5adde6.png">

I finally call this function viewDidLoad(). However, nothing appears. Am I doing something wrong? 

Have a nice day,

Sebastien.

**Comment by mousebird** [comment]
It might be in the coordinate systems.  I use several depending on what's going on.
I believe ShapeLinears use geographic in radians for (x,y) and z in display units (radius of the earth is 1.0).  Looks like Billboards are the same so that's not likely to be the problem.

Try giving this a really big DrawPriority.

**Comment by Sebboulnois** [comment]
Indeed, my billboards appear correctly on the globe. I am then using their coordinates for building the MaplyShapeLinear object. I believe my billboards are taking a value in meters (or feet?) which is 15 000. Billboards and MaplyShapeLinear objects are both using MaplyCoordinate3D properties so it should be okay, right?

I tried with a priority of 1 million but still did not appear.

What concerns me is the UnsafeMutablePointer part. Am I doing it right? I am not really familiar with this concept and I believe I didn't have to use that in Objective-C.

**Comment by mousebird** [comment]
Swift doesn't like simple arrays of objects.  That ought to be fine.

I think I see the problem.  The ShapeLinear z value is an offset from the surface of the earth, but the scale is relative to the radius of the earth (1.0 here).  The Billboards use a more logical meters from the surface of the earth offset.  So you'll need to turn one into the other.

**Comment by Sebboulnois** [comment]
I am little bit confused. Does it mean that I need to divide my value in meters by the earth radius and then add 1 to it as it follows?

![screen shot 2017-09-05 at 2 01 10 pm](https://user-images.githubusercontent.com/15698398/30075334-b447899e-9242-11e7-98ab-4354e4d2c00d.png)

Sebastien.

**Comment by stevesaildrone** [comment]
Right, I believe that's it.

**Comment by mousebird** [comment]
Right, I believe that's it.

**Comment by Sebboulnois** [comment]
Nothing appears:

![unadjustednonraw_thumb_1174](https://user-images.githubusercontent.com/15698398/30080422-90a8c660-9251-11e7-8184-496c77dc7531.jpg)

I set the priority up to different values but it doesn't seem to change anything.

**Comment by mousebird** [comment]
Not sure what to suggest.  Maybe try it with a simple well defined line and go from there.

**Comment by Sebboulnois** [comment]
Good morning,

I read again the documentation about the MaplyShapeLinear. It says: "The z values are offsets from the globe (or map) and are in display units. For the globe display units are based on a radius of 1.0.". I replaced the (value in meters) / earthRadius +1 by (value in meters) / earthRadius. I think it makes more sense. 

I then created a simple array of 2 coordinates corresponding respectively to Miami and Denver airport coordinates to make it simple and I then created a MaplyShapeLinear from this array:

<img width="815" alt="screen shot 2017-09-06 at 9 15 25 am" src="https://user-images.githubusercontent.com/15698398/30113829-fcf33944-92e3-11e7-909c-3e7c39d89d20.png">

Nothing displays, still. I don't really understand what is going on here. From here, I would have 2 questions:

1) Is my globe.add function correct?
2) Are you sure the way I manipulate the UnsafeMutablePointer object is correct?

Sebastien.

**Comment by Sebboulnois** [comment]
Good morning,

It actually works now. I was giving the UnsafeMutablepointer as the entry of the addShapes instead of the MaplyLinear itself this whole time... 

I apologize for wasting your time with this issue. Thank you!

Have a nice day,

Sebastien.
